File: Backpack/Button/Classes/BPKButtonAppearanceSets.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2025 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "BPKButtonAppearanceSets.h"

#import "BPKButtonAppearance.h"
#import "BPKButtonAppearanceSet.h"

#import <Backpack/Color.h>

NS_ASSUME_NONNULL_BEGIN

static inline BPKButtonAppearance *BPKButtonAppearanceWith(UIColor *background, UIColor *foreground) {
    return [[BPKButtonAppearance alloc] initWithBorderColor:nil
                                         gradientStartColor:background
                                           gradientEndColor:background
                                            foregroundColor:foreground];
}

@implementation BPKButtonAppearanceSets

+ (BPKButtonAppearance *)disabledAppearance {
    static BPKButtonAppearance *appearance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        appearance = BPKButtonAppearanceWith(BPKColor.buttonDisabledBackgroundColor, BPKColor.textDisabledColor);
    });
    return appearance;
}

+ (BPKButtonAppearanceSet *)primary {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryNormalBackgroundColor, BPKColor.textOnDarkColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryPressedBackgroundColor, BPKColor.textOnDarkColor)
                                                      disabledAppearance:[self disabledAppearance]
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryPressedBackgroundColor, BPKColor.textOnDarkColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)secondary {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryNormalBackgroundColor, BPKColor.textPrimaryColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryPressedBackgroundColor, BPKColor.textPrimaryColor)
                                                      disabledAppearance:[self disabledAppearance]
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryPressedBackgroundColor, BPKColor.textPrimaryColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)featured {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonFeaturedNormalBackgroundColor, BPKColor.textPrimaryInverseColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonFeaturedPressedBackgroundColor, BPKColor.textPrimaryInverseColor)
                                                      disabledAppearance:[self disabledAppearance]
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonFeaturedPressedBackgroundColor, BPKColor.textPrimaryInverseColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)destructive {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonDestructiveNormalBackgroundColor, BPKColor.buttonDestructiveNormalForegroundColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonDestructivePressedBackgroundColor, BPKColor.textPrimaryInverseColor)
                                                      disabledAppearance:[self disabledAppearance]
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonDestructivePressedBackgroundColor, BPKColor.textPrimaryInverseColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)link {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        BPKButtonAppearance *disabled = BPKButtonAppearanceWith([[UIColor clearColor] colorWithAlphaComponent:0], BPKColor.textDisabledColor);
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkNormalForegroundColor)
                                                       loadingAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkPressedForegroundColor)
                                                      disabledAppearance:disabled
                                                   highlightedAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkPressedForegroundColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)linkOnDark {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        BPKButtonAppearance *disabled = BPKButtonAppearanceWith([[UIColor clearColor] colorWithAlphaComponent:0], BPKColor.buttonLinkOnDarkDisabledForegroundColor);
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkOnDarkNormalForegroundColor)
                                                       loadingAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkOnDarkPressedForegroundColor)
                                                      disabledAppearance:disabled
                                                   highlightedAppearance:BPKButtonAppearanceWith([UIColor clearColor], BPKColor.buttonLinkOnDarkPressedForegroundColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)primaryOnDark {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnDarkNormalBackgroundColor, BPKColor.textOnLightColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnDarkPressedBackgroundColor, BPKColor.textOnLightColor)
                                                      disabledAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnDarkDisabledBackgroundColor, BPKColor.buttonPrimaryOnDarkDisabledForegroundColor)
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnDarkPressedBackgroundColor, BPKColor.textOnLightColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)primaryOnLight {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnLightNormalBackgroundColor, BPKColor.textOnDarkColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnLightPressedBackgroundColor, BPKColor.textOnDarkColor)
                                                      disabledAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnLightDisabledBackgroundColor, BPKColor.buttonPrimaryOnLightDisabledForegroundColor)
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonPrimaryOnLightPressedBackgroundColor, BPKColor.textOnDarkColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)secondaryOnDark {
    static BPKButtonAppearanceSet *set;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        set = [[BPKButtonAppearanceSet alloc] initWithRegularAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryOnDarkNormalBackgroundColor, BPKColor.textOnDarkColor)
                                                       loadingAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryOnDarkPressedBackgroundColor, BPKColor.textOnDarkColor)
                                                      disabledAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryOnDarkDisabledBackgroundColor, BPKColor.buttonSecondaryOnDarkDisabledForegroundColor)
                                                   highlightedAppearance:BPKButtonAppearanceWith(BPKColor.buttonSecondaryOnDarkPressedBackgroundColor, BPKColor.textOnDarkColor)];
    });
    return set;
}

+ (BPKButtonAppearanceSet *)appearanceFromStyle:(BPKButtonStyle)style {
    switch (style) {
    case BPKButtonStylePrimary:
        return [self primary];
    case BPKButtonStyleSecondary:
        return [self secondary];
    case BPKButtonStyleFeatured:
        return [self featured];
    case BPKButtonStyleDestructive:
        return [self destructive];
    case BPKButtonStyleLink:
        return [self link];
    case BPKButtonStyleLinkOnDark:
        return [self linkOnDark];
    case BPKButtonStylePrimaryOnDark:
        return [self primaryOnDark];
    case BPKButtonStylePrimaryOnLight:
        return [self primaryOnLight];
    case BPKButtonStyleSecondaryOnDark:
        return [self secondaryOnDark];
    default:
        return [self primary];
    }
}

@end

NS_ASSUME_NONNULL_END

File: Backpack/Button/Classes/BPKButtonAppearanceSets.swift
/*
* Backpack - Skyscanner's Design System
*
* Copyright 2018 Skyscanner Ltd
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import Foundation

@objcMembers
@IBDesignable
public class BPKButtonAppearanceSets: NSObject {
        
    private static var disabledAppearance = BPKButtonAppearance.with(
        background: BPKColor.buttonDisabledBackgroundColor,
        foreground: BPKColor.textDisabledColor
    )
    
    public static var primary = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonPrimaryNormalBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        loading: .with(
            background: BPKColor.buttonPrimaryPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        disabledAppearance: disabledAppearance,
        highlightedAppearance: .with(
            background: BPKColor.buttonPrimaryPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor
        )
    )
    
    public static var secondary = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonSecondaryNormalBackgroundColor,
            foreground: BPKColor.textPrimaryColor),
        loading: .with(
            background: BPKColor.buttonSecondaryPressedBackgroundColor,
            foreground: BPKColor.textPrimaryColor),
        disabledAppearance: disabledAppearance,
        highlightedAppearance: .with(
            background: BPKColor.buttonSecondaryPressedBackgroundColor,
            foreground: BPKColor.textPrimaryColor
        )
    )
    
    public static var featured = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonFeaturedNormalBackgroundColor,
            foreground: BPKColor.textPrimaryInverseColor),
        loading: .with(
            background: BPKColor.buttonFeaturedPressedBackgroundColor,
            foreground: BPKColor.textPrimaryInverseColor),
        disabledAppearance: disabledAppearance,
        highlightedAppearance: .with(
            background: BPKColor.buttonFeaturedPressedBackgroundColor,
            foreground: BPKColor.textPrimaryInverseColor
        )
    )
    
    public static var destructive = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonDestructiveNormalBackgroundColor,
            foreground: BPKColor.buttonDestructiveNormalForegroundColor),
        loading: .with(
            background: BPKColor.buttonDestructivePressedBackgroundColor,
            foreground: BPKColor.textPrimaryInverseColor),
        disabledAppearance: disabledAppearance,
        highlightedAppearance: .with(
            background: BPKColor.buttonDestructivePressedBackgroundColor,
            foreground: BPKColor.textPrimaryInverseColor
        )
    )
    
    public static var link = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkNormalForegroundColor),
        loading: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkPressedForegroundColor),
        disabledAppearance: .with(
            background: BPKColor.clear.withAlphaComponent(0),
            foreground: BPKColor.textDisabledColor
        ),
        highlightedAppearance: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkPressedForegroundColor
        )
    )
    
    public static var linkOnDark = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkOnDarkNormalForegroundColor),
        loading: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkOnDarkPressedForegroundColor),
        disabledAppearance: .with(
            background: BPKColor.clear.withAlphaComponent(0),
            foreground: BPKColor.buttonLinkOnDarkDisabledForegroundColor
        ),
        highlightedAppearance: .with(
            background: BPKColor.clear,
            foreground: BPKColor.buttonLinkOnDarkPressedForegroundColor
        )
    )
    
    public static var primaryOnDark = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonPrimaryOnDarkNormalBackgroundColor,
            foreground: BPKColor.textOnLightColor),
        loading: .with(
            background: BPKColor.buttonPrimaryOnDarkPressedBackgroundColor,
            foreground: BPKColor.textOnLightColor),
        disabledAppearance: .with(
            background: BPKColor.buttonPrimaryOnDarkDisabledBackgroundColor,
            foreground: BPKColor.buttonPrimaryOnDarkDisabledForegroundColor
        ),
        highlightedAppearance: .with(
            background: BPKColor.buttonPrimaryOnDarkPressedBackgroundColor,
            foreground: BPKColor.textOnLightColor
        )
    )
    
    public static var secondaryOnDark = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonSecondaryOnDarkNormalBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        loading: .with(
            background: BPKColor.buttonSecondaryOnDarkPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        disabledAppearance: .with(
            background: BPKColor.buttonSecondaryOnDarkDisabledBackgroundColor,
            foreground: BPKColor.buttonSecondaryOnDarkDisabledForegroundColor
        ),
        highlightedAppearance: .with(
            background: BPKColor.buttonSecondaryOnDarkPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor
        )
    )
    
    public static var primaryOnLight = BPKButtonAppearanceSet(
        regularAppearance: .with(
            background: BPKColor.buttonPrimaryOnLightNormalBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        loading: .with(
            background: BPKColor.buttonPrimaryOnLightPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor),
        disabledAppearance: .with(
            background: BPKColor.buttonPrimaryOnLightDisabledBackgroundColor,
            foreground: BPKColor.buttonPrimaryOnLightDisabledForegroundColor
        ),
        highlightedAppearance: .with(
            background: BPKColor.buttonPrimaryOnLightPressedBackgroundColor,
            foreground: BPKColor.textOnDarkColor
        )

    )
    
    public static func appearance(fromStyle style: BPKButtonStyle) -> BPKButtonAppearanceSet {
        switch style {
        case .primary: return self.primary
        case .secondary: return self.secondary
        case .secondaryOnDark: return self.secondaryOnDark
        case .destructive: return self.destructive
        case .featured: return self.featured
        case .link: return self.link
        case .linkOnDark: return self.linkOnDark
        case .primaryOnDark: return self.primaryOnDark
        case .primaryOnLight: return self.primaryOnLight
        default: return self.primary
        }
    }
}

fileprivate extension UIColor {
    static var pressedMonteverde: UIColor {
        UIColor(red: 0, green: 0.416, blue: 0.38, alpha: 1)
    }
    
    static var clear = BPKColor.clear
    
    static var primaryColor = BPKColor.skyBlue.darkVariant(.bpk.skyBlueTint01)
    
    static var destructiveColor: UIColor {
        UIColor(red: 0.699, green: 0.182, blue: 0.269, alpha: 1)
            .darkVariant(
                UIColor(red: 0.972, green: 0.361, blue: 0.465, alpha: 1)
            )
    }
    
    static var bpk = BPKColor.self
    
    func darkVariant(_ dark: UIColor) -> UIColor {
        BPKColor.dynamicColor(withLightVariant: self, darkVariant: dark)
    }
}

fileprivate extension BPKButtonAppearance {
    static func with(background: UIColor, foreground: UIColor) -> BPKButtonAppearance {
        BPKButtonAppearance(
            borderColor: nil,
            gradientStart: background,
            gradientEnd: background,
            foregroundColor: foreground
        )
    }
}
